core/system/src/mpu/vmpu_freescale_k64_mem.c

```diff
@@ -81,24 +143,33 @@ void vmpu_mem_switch(uint8_t src_box, uint8_t dst_box)
     rgd = box->acl;
     assert(rgd);
 
-    /* update MPU regions */
-    t = g_mem_acl_user + 1;
-    u = dst_count;
-    while(u--)
-    {
-        word = MPU->WORD[t++];
+    /* Update MPU regions:
+     * We assume that the first two ACLs belonging to the box are the stack
+     * and context ACLs. Therefore we start at index K64F_MPU_REGIONS_RESERVED.
+     * We opportunistically copy all remaining ACLs into the MPU regions, until
+     * the regions run out. The remaining ACLs are switched in on demand. */
+    u = K64F_MPU_REGIONS_RESERVED + dst_count;
+    /* The round robin _starts_ at the first free MPU region _after_ the box
+     * ACLs have been copied. */
+    g_mpu_slot = u;
+    if (u > K64F_MPU_REGIONS_MAX) {
+        /* Clamp u to K64F_MPU_REGIONS_MAX. */
+        u = K64F_MPU_REGIONS_MAX;
+        /* If no MPU regions are free anymore, set the RR marker at the
+         * of the USER MPU regions (leaving stack and context intact!). */
+        g_mpu_slot = K64F_MPU_REGIONS_USER;
+    } else {
+        /* Copy the active page heap ACLs into the MPU regions:
+         * This will start at index `g_mpu_slot` and continue to `K64F_MPU_REGIONS_MAX`. */
+        page_allocator_iterate_active_pages(vmpu_mem_push_page_acl_iterator);
+    }
+    /* Copy the ACLs into the MPU regions. */
+    for (t = K64F_MPU_REGIONS_RESERVED; t < u; t++, rgd++) {
+        word = MPU->WORD[t];
```

Comment:
@niklas-arm instead of the array-assignment - in case there's not a struct for a single MPU slot provided by Freescale, please create one and use names instead of array indexes (with volatile members etc.)